pkg/block/block.go

```diff
@@ -126,6 +129,35 @@ func cleanUp(logger log.Logger, bkt objstore.Bucket, id ulid.ULID, err error) er
 	return err
 }
 
+// MarkForDeletion creates a file which stores information about when the block was marked for deletion.
+func MarkForDeletion(ctx context.Context, logger log.Logger, bkt objstore.Bucket, id ulid.ULID, blocksMarkedForDeletion prometheus.Counter) error {
+	deletionMarkFile := path.Join(id.String(), metadata.DeletionMarkFilename)
+	deletionMarkExists, err := objstore.Exists(ctx, bkt, deletionMarkFile)
+	if err != nil {
+		return errors.Wrapf(err, "check exists: %s in bucket", deletionMarkFile)
+	}
+	if deletionMarkExists {
+		return errors.Wrapf(err, "file %s already exists in bucket", deletionMarkFile)
```

Comment:
I don't think we need to return error from here. If file already exists and block is marked for deletion, there is nothing more to do. (Perhaps checking if it is a correct marker, but that sounds like overkill).

Reply:
Bartek pointed out that we should return an error in this case. I am of the opinion that a block should be marked for deletion once and if it is marked again, we might have race condition which we are missing.

Reply:
I think from the caller's point of view, what it wants to do (mark block as deleted) is already done. That's why I think that return nil (and perhaps a warning log message?) would be enough. Otherwise, caller will see error and return error to its called, but for no good reason.

Reply:
Even though we don't have it explicitly, The caller expects to mark the block as deleted with a block deletion timestamp.
In the case that the block was already marked as deleted, we will be having an older timestamp about when the block was marked to be deleted.

Reply:
Agree with @khyatisoneji 
This is the question how you deal with 404. I am in favor of erroring out. This way you can detect bugs early.

Reply:
OK, if it is expected to only call this once for a block, then I can understand this reasoning.

Reply:
I am marking this as resolved, feel free to re-open this conversation.